### ISSUE TITLE
EGLNOPERMISSION and EINVALIDGLTOKEN thrown even with a valid token that has the correct permissions

### ISSUE THREAD

**Issue by esciara** [semantic-release/gitlab#92]
Despite having set `GITLAB_TOKEN` with a valid token that has the correct permissions, I have the following error:
```
[8:33:25 AM] [semantic-release] [@semantic-release/gitlab] › ℹ  Verify GitLab authentication (https://gitlab.com/api/v4)
[8:33:26 AM] [semantic-release] › ✖  Failed step "verifyConditions" of plugin "@semantic-release/gitlab"
[8:33:26 AM] [semantic-release] › ✖  EGLNOPERMISSION The GitLab token doesn't allow to push on the repository root/my-project.
The user associated with the GitLab token (https://github.com/semantic-release/gitlab/blob/master/README.md#gitlab-authentication) configured in the GL_TOKEN or GITLAB_TOKEN environment variable must allows to push to the repository root/my-project.

Please make sure the GitLab user associated with the token has the permission to push (https://docs.gitlab.com/ee/user/permissions.html#project-members-permissions) to the repository root/my-project.

[8:33:26 AM] [semantic-release] › ✖  EINVALIDGLTOKEN Invalid GitLab token.
The GitLab token (https://github.com/semantic-release/gitlab/blob/master/README.md#gitlab-authentication) configured in the GL_TOKEN or GITLAB_TOKEN environment variable must be a valid personal access token (https://docs.gitlab.com/ce/user/profile/personal_access_tokens.html) allowing to push to the repository root/my-project.

Please make sure to set the GL_TOKEN or GITLAB_TOKEN environment variable in your CI with the exact value of the GitLab personal token.

AggregateError: 
    SemanticReleaseError: Invalid GitLab token.
        at module.exports (/usr/local/lib/node_modules/@semantic-release/gitlab/lib/get-error.js:6:10)
        at module.exports (/usr/local/lib/node_modules/@semantic-release/gitlab/lib/verify.js:58:21)
    SemanticReleaseError: The GitLab token doesn't allow to push on the repository root/my-project.
        at module.exports (/usr/local/lib/node_modules/@semantic-release/gitlab/lib/get-error.js:6:10)
        at module.exports (/usr/local/lib/node_modules/@semantic-release/gitlab/lib/verify.js:67:19)
    at /usr/local/lib/node_modules/semantic-release/lib/plugins/pipeline.js:54:11
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Object.pluginsConf.<computed> [as verifyConditions] (/usr/local/lib/node_modules/semantic-release/lib/plugins/index.js:80:11)
    at async run (/usr/local/lib/node_modules/semantic-release/index.js:84:3)
    at async module.exports (/usr/local/lib/node_modules/semantic-release/index.js:170:22)
    at async module.exports (/usr/local/lib/node_modules/semantic-release/cli.js:55:5) {
  name: 'AggregateError'
}
```

My project has the id 3 on gitlab, I receive the correct response when doing the following `curl`:
```
$ curl --header "Private-Token: ${GITLAB_TOKEN}" ${GITLAB_URL}${GITLAB_PREFIX}/projects/3
{"id":3,"description":"","name":"my-project", ... }
```

However when doing a curl using the [project's namespace](https://docs.gitlab.com/ee/api/README.html#namespaced-path-encoding) (as does the plugin's code), I have the following response:
```
$ curl --header "Private-Token: ${GITLAB_TOKEN}" ${GITLAB_URL}${GITLAB_PREFIX}/projects/root%2Fmy-project
{"error":"404 Not Found"}
```

So here are several issues:
 1. The plugin should not display both `EGLNOPERMISSION` and `EINVALIDGLTOKEN`. Either the token is valid and has not the right permissions, or it is invalid and being invalid it cannot have any permission anyway.
 2. Some GitLab instances give a proper response when using the ID of the project, but a 404 error when using the project's namespace. The 404 is badly interpreted by the plugin.

**Comment by esciara** [comment]
After doing some research around the issue with using project's namespace to find gitlab projects, I re-discovered the details of problematic reverse-proxy configurations in https://github.com/semantic-release/gitlab/issues/63#issuecomment-527080331 (I actually wrote this comment and forgot about it...):

> As another note, if you still cannot access your project with this `curl`, try `curl --header "Private-Token: <your_access_token>" https://gitlab.com/api/v4/projects`. If this time you get a response, if you also are behind a reverse proxy, it might be that your reverse proxy is not correctly configured, and decodes the `%2F` into a `/`, hence forwarding a `https://gitlab.com/api/v4/projects/mygroup/myrepo`, which will result in a `{"error": "404 Not Found"}`.
> 
> See [this GitLab issue](https://gitlab.com/gitlab-org/gitlab-ce/issues/35079) for more details, including a correction of reverse proxy configuration. Careful about the solution provided, it might not be a full proof secure solution of Nginx revers proxy, see [this Nginx ticket](https://trac.nginx.org/nginx/ticket/727) for more info and [this stackoverflow entry](https://stackoverflow.com/a/37584637) for a potentially more secure solution.

I have found other reports the web of this being an issue, including one specific to `semantic-release` at https://github.com/semantic-release/semantic-release/issues/1073 and others on the web. This is an issue currently in my company. Since the error massages are not very helpful, I am wondering how many other people had the same issue, and abandoned the usage of `semantic-release/gitlab` as a result (I certainly nearly did).

**Comment by pvdlg** [comment]
Indeed we shouldn't send both `EGLNOPERMISSION` and `EINVALIDGLTOKEN`. See fix in https://github.com/semantic-release/gitlab/pull/93

However I don't think the a `404` response cause this. If the GitLab API was responding with a `404` [the plugin would throw a `EMISSINGREPO` error](https://github.com/semantic-release/gitlab/blob/4a1059e92c5c298679b9f18d849923b349b45a8e/lib/verify.js#L60).
Because it throws a `EINVALIDGLTOKEN` the GitLab API must have responded with a `401`.

**Comment by esciara** [comment]
Here is what is going through my pipeline:
```
$ echo $CI_COMMIT_REF_NAME
test
$ echo $CI_PROJECT_URL
http://gitlab.example.com/root/my-project
$ echo $CI_PROJECT_PATH
root/my-project
$ echo $CI_API_V4_URL
http://gitlab.example.com/api/v4
$ GITLAB_URL=${CI_PROJECT_URL%/$CI_PROJECT_PATH}
$ echo $GITLAB_URL
http://gitlab.example.com
$ GITLAB_PREFIX=${CI_API_V4_URL#$GITLAB_URL}
$ echo $GITLAB_PREFIX
/api/v4
$ echo $GITLAB_TOKEN
erx5r5MxqecZFHp8PrBy
$ curl --header "Private-Token: ${GITLAB_TOKEN}" ${GITLAB_URL}${GITLAB_PREFIX}/projects
{"id":3,"description":"","name":"my-project", ... }
$ curl --header "Private-Token: ${GITLAB_TOKEN}" ${GITLAB_URL}${GITLAB_PREFIX}/projects/root%2Fmy-project
{"error":"404 Not Found"}
$ semantic-release --branch $CI_COMMIT_REF_NAME --repository-url $CI_PROJECT_URL
[..]
[10:17:52 PM] [semantic-release] [@semantic-release/gitlab] › ℹ  Verify GitLab authentication (https://gitlab.com/api/v4)
[10:17:52 PM] [semantic-release] › ✖  Failed step "verifyConditions" of plugin "@semantic-release/gitlab"
[10:17:52 PM] [semantic-release] › ✖  EINVALIDGLTOKEN Invalid GitLab token.
The GitLab token (https://github.com/semantic-release/gitlab/blob/master/README.md#gitlab-authentication) configured in the GL_TOKEN or GITLAB_TOKEN environment variable must be a valid personal access token (https://docs.gitlab.com/ce/user/profile/personal_access_tokens.html) allowing to push to the repository root/my-project.

Please make sure to set the GL_TOKEN or GITLAB_TOKEN environment variable in your CI with the exact value of the GitLab personal token.

AggregateError:
    SemanticReleaseError: Invalid GitLab token.
        at module.exports (/usr/local/lib/node_modules/@semantic-release/gitlab/lib/get-error.js:6:10)
        at module.exports (/usr/local/lib/node_modules/@semantic-release/gitlab/lib/verify.js:62:21)
    at /usr/local/lib/node_modules/semantic-release/lib/plugins/pipeline.js:54:11
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Object.pluginsConf.<computed> [as verifyConditions] (/usr/local/lib/node_modules/semantic-release/lib/plugins/index.js:80:11)
    at async run (/usr/local/lib/node_modules/semantic-release/index.js:84:3)
    at async module.exports (/usr/local/lib/node_modules/semantic-release/index.js:170:22)
    at async module.exports (/usr/local/lib/node_modules/semantic-release/cli.js:55:5) {
  name: 'AggregateError'
}
```

And if I look into that code:

https://github.com/semantic-release/gitlab/blob/1cc5f441432fa69aa97250377d2d73918593823a/lib/verify.js#L47-L68

It means that the error is thrown when calling:
```javascript
      got.get(urlJoin(apiUrl, `/projects/${encodeURIComponent(repoId)}`), {
        json: true,
        headers: {'Private-Token': gitlabToken},
      })
```

And the equivalent for ``urlJoin(apiUrl, `/projects/${encodeURIComponent(repoId)}`)`` should be `${GITLAB_URL}${GITLAB_PREFIX}/projects/root%2Fmy-project`. But as it shows on my pipeline, I do not get an error `401` by curling on this but a `404`:
```
${GITLAB_URL}${GITLAB_PREFIX}/projects/root%2Fmy-project
{"error":"404 Not Found"}
```

Am I thinking wrong here?

Petty I don't know how to run that in debug mode... a walkthrough would have helped. If anyonw has a suggestion...

**Comment by pvdlg** [comment]
I don't know...It seems pretty clear that what you get is a `401` because that's the only way the plugin can return the error `EINVALIDGLTOKEN`.
I don't know how to debug that with your particular environment. Maybe you can fork this repo and add some `console.log` to try to figure out what's going on.

**Comment by esciara** [comment]
OK, I will be digging. In the meantime, I have worked on improving the tests so that the pb of `EGLNOPERMISSION` and `EINVALIDGLTOKEN` being thrown would be caught. I am discovering some other things in the process. Will open another issue for this particular work on tests improvements.

**Comment by esciara** [comment]
I have found the origin of the `Invalid GitLab token`: I needed to add `export` to propagate the environment variables to `semantic-release`, hence:
```
$ export GITLAB_URL=${CI_PROJECT_URL%/$CI_PROJECT_PATH}
$ export GITLAB_PREFIX=${CI_API_V4_URL#$GITLAB_URL}
```

When missing `export`, then the `gitlabUrl` defaults to `https://gitlab.com`... and indeed my token is not valid there. I did not have the problem with `GITLAB_TOKEN` because it was set though the [GitLab interface](https://docs.gitlab.com/ee/ci/variables/#custom-environment-variables) for use in my pipeline.

Actually, one needs to be really careful in reading the outputs to be able to see the issue... It does not jump to one's eyes when it happens.

I will be writing a PR with more debug output and more contextual info in the error message to help with debugging. 

I will also add another check on the output of the 404 response from GitLab: the code currently does not make a distinction when the repository is really missing (`statusMessage` is `Project Not Found`) and when there is a problem with a badly configured reverse-proxy (`statusMessage` is `Not Found`).
